File: web/src/crypto/keys.test.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Serraniel and the Sendan contributors

import { describe, expect, it } from "vitest";
import { KeyMaterialError } from "./errors.js";
import {
  authTokenHash,
  deriveKeys,
  deriveKeysWithPassword,
  FILE_ID_SIZE,
  LINK_SECRET_SIZE,
  newLinkSecret,
  newOwnerToken,
  ownerTokenHash,
} from "./keys.js";
import { newPasswordParams, PASSWORD_SALT_SIZE, type PasswordParams } from "./password.js";

const fixedFileID = () => new Uint8Array(FILE_ID_SIZE).fill(0x01);
const fixedLinkSecret = () => new Uint8Array(LINK_SECRET_SIZE).fill(0x02);

/** Deliberately weak so tests stay fast. Never use these for real. */
const fixedParams = (): PasswordParams => ({
  salt: new Uint8Array(PASSWORD_SALT_SIZE).fill(0x03),
  memoryKiB: 64,
  iterations: 1,
  parallelism: 1,
});

const hex = (b: Uint8Array) => Buffer.from(b).toString("hex");

describe("key schedule", () => {
  it("is deterministic", async () => {
    const a = await deriveKeys(fixedFileID(), fixedLinkSecret());
    const b = await deriveKeys(fixedFileID(), fixedLinkSecret());
    expect(hex(a.wrapping)).toBe(hex(b.wrapping));
    expect(hex(a.metadata)).toBe(hex(b.metadata));
    expect(hex(a.authToken)).toBe(hex(b.authToken));
  });

  // If a label were duplicated or dropped, two keys would collide and
  // compromising one would compromise another.
  it("derives three distinct 32-byte keys", async () => {
    const k = await deriveKeys(fixedFileID(), fixedLinkSecret());
    expect(new Set([hex(k.wrapping), hex(k.metadata), hex(k.authToken)]).size).toBe(3);
    for (const key of [k.wrapping, k.metadata, k.authToken]) {
      expect(key.length).toBe(32);
    }
  });

  it("varies with the file id and the link secret", async () => {
    const base = await deriveKeys(fixedFileID(), fixedLinkSecret());

    const otherID = fixedFileID();
    otherID[0] = (otherID[0] ?? 0) ^ 0xff;
    expect(hex((await deriveKeys(otherID, fixedLinkSecret())).wrapping)).not.toBe(
      hex(base.wrapping),
    );

    const otherSecret = fixedLinkSecret();
    otherSecret[0] = (otherSecret[0] ?? 0) ^ 0xff;
    expect(hex((await deriveKeys(fixedFileID(), otherSecret)).wrapping)).not.toBe(
      hex(base.wrapping),
    );
  });

  it("rejects malformed inputs", async () => {
    const cases: Array<[string, Uint8Array, Uint8Array]> = [
      ["short file id", new Uint8Array(FILE_ID_SIZE - 1), fixedLinkSecret()],
      ["long file id", new Uint8Array(FILE_ID_SIZE + 1), fixedLinkSecret()],
      ["short link secret", fixedFileID(), new Uint8Array(LINK_SECRET_SIZE - 1)],
      ["16 byte link secret", fixedFileID(), new Uint8Array(16)],
    ];
    for (const [name, fileID, linkSecret] of cases) {
      await expect(deriveKeys(fileID, linkSecret), name).rejects.toBeInstanceOf(KeyMaterialError);
    }
  });
});

describe("password contribution", () => {
  // A password must change the wrapping key itself. If it did not, it would be
  // server-enforced policy rather than a cryptographic property, which is the
  // weakness this scheme exists to avoid.
  it("changes the wrapping and metadata keys", async () => {
    const without = await deriveKeys(fixedFileID(), fixedLinkSecret());
    const with_ = await deriveKeysWithPassword(
      fixedFileID(),
      fixedLinkSecret(),
      "correct horse",
      fixedParams(),
    );
    expect(hex(with_.wrapping)).not.toBe(hex(without.wrapping));
    // Without this, filenames would leak from a password-protected upload.
    expect(hex(with_.metadata)).not.toBe(hex(without.metadata));
  });

  it("distinguishes different passwords", async () => {
    const a = await deriveKeysWithPassword(fixedFileID(), fixedLinkSecret(), "a", fixedParams());
    const b = await deriveKeysWithPassword(fixedFileID(), fixedLinkSecret(), "b", fixedParams());
    expect(hex(a.wrapping)).not.toBe(hex(b.wrapping));
  });

  // An empty password denotes a meaningless state: an upload marked
  // password-protected that any link holder can open. Both implementations
  // reject it, so the state cannot arise at all.
  it("rejects an empty password", async () => {
    await expect(
      deriveKeysWithPassword(fixedFileID(), fixedLinkSecret(), "", fixedParams()),
    ).rejects.toBeInstanceOf(KeyMaterialError);
  });

  it("rejects bad parameters", async () => {
    const mutations: Array<[string, (p: PasswordParams) => void]> = [
      ["short salt", (p) => (p.salt = new Uint8Array(PASSWORD_SALT_SIZE - 1))],
      ["zero memory", (p) => (p.memoryKiB = 0)],
      ["zero iterations", (p) => (p.iterations = 0)],
      ["zero parallelism", (p) => (p.parallelism = 0)],
    ];
    for (const [name, mutate] of mutations) {
      const p = fixedParams();
      mutate(p);
      await expect(
        deriveKeysWithPassword(fixedFileID(), fixedLinkSecret(), "pw", p),
        name,
      ).rejects.toBeInstanceOf(KeyMaterialError);
    }
  });
});

describe("random values and hashes", () => {
  it("produces the specified sizes and does not repeat", () => {
    expect(newLinkSecret().length).toBe(LINK_SECRET_SIZE);
    expect(hex(newLinkSecret())).not.toBe(hex(newLinkSecret()));
    expect(hex(newOwnerToken())).not.toBe(hex(newOwnerToken()));
  });

  it("uses the spec default Argon2id parameters", () => {
    const p = newPasswordParams();
    expect(p.salt.length).toBe(PASSWORD_SALT_SIZE);
    expect(p.memoryKiB).toBe(65536);
    expect(p.iterations).toBe(3);
    expect(p.parallelism).toBe(1);
    expect(hex(p.salt)).not.toBe(hex(newPasswordParams().salt));
  });

  it("hashes tokens stably to 32 bytes", async () => {
    const token = new Uint8Array(32).fill(0x07);
    expect(hex(await authTokenHash(token))).toBe(hex(await authTokenHash(token)));
    expect((await authTokenHash(token)).length).toBe(32);
    expect((await ownerTokenHash(token)).length).toBe(32);
  });
});

File: internal/crypto/keys_test.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Serraniel and the Sendan contributors

package crypto

import (
	"bytes"
	"errors"
	"testing"
)

func fixedFileID() []byte     { return bytes.Repeat([]byte{0x01}, FileIDSize) }
func fixedLinkSecret() []byte { return bytes.Repeat([]byte{0x02}, LinkSecretSize) }

func fixedPasswordParams() PasswordParams {
	return PasswordParams{
		Salt: bytes.Repeat([]byte{0x03}, PasswordSaltSize),
		// Deliberately weak so tests stay fast. Never use these for real.
		MemoryKiB:   64,
		Iterations:  1,
		Parallelism: 1,
	}
}

func TestDeriveKeysIsDeterministic(t *testing.T) {
	a, err := DeriveKeys(fixedFileID(), fixedLinkSecret())
	if err != nil {
		t.Fatalf("first derivation: %v", err)
	}
	b, err := DeriveKeys(fixedFileID(), fixedLinkSecret())
	if err != nil {
		t.Fatalf("second derivation: %v", err)
	}
	if !bytes.Equal(a.Wrapping, b.Wrapping) || !bytes.Equal(a.Metadata, b.Metadata) || !bytes.Equal(a.AuthToken, b.AuthToken) {
		t.Fatal("same inputs produced different keys")
	}
}

// The three keys must be independent. If a label were duplicated or dropped,
// two of them would collide and compromising one would compromise another.
func TestDerivedKeysAreDistinct(t *testing.T) {
	k, err := DeriveKeys(fixedFileID(), fixedLinkSecret())
	if err != nil {
		t.Fatalf("derive: %v", err)
	}
	for _, c := range []struct {
		name string
		a, b []byte
	}{
		{"wrapping vs metadata", k.Wrapping, k.Metadata},
		{"wrapping vs auth", k.Wrapping, k.AuthToken},
		{"metadata vs auth", k.Metadata, k.AuthToken},
	} {
		if bytes.Equal(c.a, c.b) {
			t.Errorf("%s: keys are identical, domain separation has failed", c.name)
		}
	}
	for _, c := range []struct {
		name string
		key  []byte
	}{{"wrapping", k.Wrapping}, {"metadata", k.Metadata}, {"auth", k.AuthToken}} {
		if len(c.key) != derivedKeySize {
			t.Errorf("%s key is %d bytes, want %d", c.name, len(c.key), derivedKeySize)
		}
	}
}

func TestDeriveKeysVariesWithEveryInput(t *testing.T) {
	base, err := DeriveKeys(fixedFileID(), fixedLinkSecret())
	if err != nil {
		t.Fatalf("derive: %v", err)
	}

	otherID := fixedFileID()
	otherID[0] ^= 0xFF
	byID, err := DeriveKeys(otherID, fixedLinkSecret())
	if err != nil {
		t.Fatalf("derive with other file id: %v", err)
	}
	if bytes.Equal(base.Wrapping, byID.Wrapping) {
		t.Error("file id does not affect the derived keys")
	}

	otherSecret := fixedLinkSecret()
	otherSecret[0] ^= 0xFF
	bySecret, err := DeriveKeys(fixedFileID(), otherSecret)
	if err != nil {
		t.Fatalf("derive with other link secret: %v", err)
	}
	if bytes.Equal(base.Wrapping, bySecret.Wrapping) {
		t.Error("link secret does not affect the derived keys")
	}
}

// A password must change the wrapping key itself. If it did not, the password
// would be server-enforced policy rather than a cryptographic property, which
// is precisely the weakness this scheme exists to avoid.
func TestPasswordChangesTheWrappingKey(t *testing.T) {
	without, err := DeriveKeys(fixedFileID(), fixedLinkSecret())
	if err != nil {
		t.Fatalf("derive without password: %v", err)
	}
	with, err := DeriveKeysWithPassword(fixedFileID(), fixedLinkSecret(), "correct horse", fixedPasswordParams())
	if err != nil {
		t.Fatalf("derive with password: %v", err)
	}
	if bytes.Equal(without.Wrapping, with.Wrapping) {
		t.Fatal("password did not affect the wrapping key")
	}
	if bytes.Equal(without.Metadata, with.Metadata) {
		t.Fatal("password did not affect the metadata key: filenames would leak without it")
	}

	other, err := DeriveKeysWithPassword(fixedFileID(), fixedLinkSecret(), "correct horss", fixedPasswordParams())
	if err != nil {
		t.Fatalf("derive with other password: %v", err)
	}
	if bytes.Equal(with.Wrapping, other.Wrapping) {
		t.Fatal("different passwords produced the same wrapping key")
	}
}

// An empty password denotes a meaningless state: an upload marked
// password-protected that any link holder can open. The browser's Argon2id
// implementation refuses it outright, so both implementations reject it and the
// state cannot arise at all.
func TestEmptyPasswordIsRejected(t *testing.T) {
	if _, err := DeriveKeysWithPassword(fixedFileID(), fixedLinkSecret(), "", fixedPasswordParams()); !errors.Is(err, ErrKeyMaterial) {
		t.Fatalf("got %v, want ErrKeyMaterial", err)
	}
}

func TestDeriveKeysRejectsMalformedInput(t *testing.T) {
	for _, tc := range []struct {
		name       string
		fileID     []byte
		linkSecret []byte
	}{
		{"short file id", make([]byte, FileIDSize-1), fixedLinkSecret()},
		{"long file id", make([]byte, FileIDSize+1), fixedLinkSecret()},
		{"nil file id", nil, fixedLinkSecret()},
		{"short link secret", fixedFileID(), make([]byte, LinkSecretSize-1)},
		{"16 byte link secret", fixedFileID(), make([]byte, 16)},
		{"nil link secret", fixedFileID(), nil},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := DeriveKeys(tc.fileID, tc.linkSecret); !errors.Is(err, ErrKeyMaterial) {
				t.Fatalf("got %v, want ErrKeyMaterial", err)
			}
		})
	}
}

func TestDeriveKeysWithPasswordRejectsBadParams(t *testing.T) {
	for _, tc := range []struct {
		name   string
		mutate func(*PasswordParams)
	}{
		{"short salt", func(p *PasswordParams) { p.Salt = make([]byte, PasswordSaltSize-1) }},
		{"nil salt", func(p *PasswordParams) { p.Salt = nil }},
		{"zero memory", func(p *PasswordParams) { p.MemoryKiB = 0 }},
		{"zero iterations", func(p *PasswordParams) { p.Iterations = 0 }},
		{"zero parallelism", func(p *PasswordParams) { p.Parallelism = 0 }},
	} {
		t.Run(tc.name, func(t *testing.T) {
			p := fixedPasswordParams()
			tc.mutate(&p)
			if _, err := DeriveKeysWithPassword(fixedFileID(), fixedLinkSecret(), "pw", p); !errors.Is(err, ErrKeyMaterial) {
				t.Fatalf("got %v, want ErrKeyMaterial", err)
			}
		})
	}
}

func TestNewPasswordParamsUsesSpecDefaults(t *testing.T) {
	p, err := NewPasswordParams()
	if err != nil {
		t.Fatalf("new params: %v", err)
	}
	if len(p.Salt) != PasswordSaltSize {
		t.Errorf("salt is %d bytes, want %d", len(p.Salt), PasswordSaltSize)
	}
	if p.MemoryKiB != DefaultMemoryKiB || p.Iterations != DefaultIterations || p.Parallelism != DefaultParallelism {
		t.Errorf("got m=%d t=%d p=%d, want m=%d t=%d p=%d",
			p.MemoryKiB, p.Iterations, p.Parallelism,
			DefaultMemoryKiB, DefaultIterations, DefaultParallelism)
	}

	other, err := NewPasswordParams()
	if err != nil {
		t.Fatalf("new params: %v", err)
	}
	if bytes.Equal(p.Salt, other.Salt) {
		t.Fatal("two calls produced the same salt")
	}
}

func TestRandomGeneratorsProduceCorrectSizes(t *testing.T) {
	for _, tc := range []struct {
		name string
		gen  func() ([]byte, error)
		want int
	}{
		{"file id", NewFileID, FileIDSize},
		{"link secret", NewLinkSecret, LinkSecretSize},
		{"file key", NewFileKey, FileKeySize},
		{"owner token", NewOwnerToken, OwnerTokenSize},
	} {
		t.Run(tc.name, func(t *testing.T) {
			a, err := tc.gen()
			if err != nil {
				t.Fatalf("generate: %v", err)
			}
			if len(a) != tc.want {
				t.Fatalf("got %d bytes, want %d", len(a), tc.want)
			}
			b, err := tc.gen()
			if err != nil {
				t.Fatalf("generate: %v", err)
			}
			if bytes.Equal(a, b) {
				t.Fatal("two calls returned identical bytes")
			}
		})
	}
}

func TestTokenHashesAreStableAndSized(t *testing.T) {
	token := bytes.Repeat([]byte{0x07}, OwnerTokenSize)
	if got := AuthTokenHash(token); len(got) != 32 || !bytes.Equal(got, AuthTokenHash(token)) {
		t.Error("auth token hash is not stable or not 32 bytes")
	}
	if got := OwnerTokenHash(token); len(got) != 32 || !bytes.Equal(got, OwnerTokenHash(token)) {
		t.Error("owner token hash is not stable or not 32 bytes")
	}
}
